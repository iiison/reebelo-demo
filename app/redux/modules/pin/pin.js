import store from '$CONFIGS/store'
import { post } from '$UTILS/requestHandler'
import {
  makeReducer,
  makeActions
} from '$RUTILS/reduxUtils'

const SET_PIN = 'SET_PIN'
const UPDATE_TRIES_COUNT = 'UPDATE_TRIES_COUNT'
const {
  defaultAction: setPin,
  successAction: setPinSuccess,
  failureAction: setPinFailure,
} = makeActions(SET_PIN)

// Sync Action Creator
export function updateTriesCount() {
  return {
    type: UPDATE_TRIES_COUNT
  }
}
// **********************************************
function makePayloadForAndroid(response) {
  if (!response) {
    return
  }

  const shellResponse = {
    data: {
      webViewResponse: {
        token: undefined,
        showWebView: false
      },
      webViewError: {
        errorCode: undefined
      }
    }
  }
  const { status, data } = response

  if (status) {
    shellResponse.data.webViewError.errorCode = status.respCode
  }

  if (data) {
    shellResponse.data.webViewResponse.token = data.idToken
  }

  return shellResponse
}

function sendDataToAndroid(status, data) {
  const stringifiedData = JSON.stringify(data)

  if (window.VisaSafeClickHandler && window.VisaSafeClickHandler.vscPinWebviewResponseHandler) {
    window.VisaSafeClickHandler.vscPinWebviewResponseHandler(status, stringifiedData)
  }
}

export function addPVVToPayload() {
  const { configs : { pvv } } = store.getState()

  if (window.VisaSafeClickHandler && window.VisaSafeClickHandler.encryptPVV) {
    return {
      pvv : window.VisaSafeClickHandler.encryptPVV(pvv)
    }
  }

  return {}
}

// Async Action Creators Starts
export function triggerSetPin({
  pin,
  deviceId,
  previousPin,
  accountReferenceId,
  additionalPayLoad = {}
}) {
  return async (dispatch) => {
    dispatch(setPin())

    try {
      const payload = {
        pin,
        accountReferenceId,
        ...additionalPayLoad,
        ...addPVVToPayload(),
        confirmPin : previousPin,
        apiAction  : 'SUBMIT'
      }

      if (pin === undefined) {
        payload.apiAction = 'CANCEL'
      }

      const response = await post({
        payload,
        path: `v1/devices/${deviceId}/pin/set`,
      })

      if (response.status && response.status.respCode) {
        dispatch(setPinFailure(response.status.respCode))
      } else {
        dispatch(setPinSuccess(response))
      }

      sendDataToAndroid(response.httpStatus, makePayloadForAndroid(response))
    } catch (error) {
      dispatch(setPinFailure(error.message || error))
    }
  }
}

export function triggerVerifyPin({
  pin,
  deviceId,
  accountReferenceId,
  additionalPayLoad = {}
}) {
  const verifyPin = setPin
  const verifyPinSuccess = setPinSuccess
  const verifyPinFailure = setPinFailure

  return async (dispatch) => {
    dispatch(verifyPin())

    try {
      const payload = {
        pin,
        accountReferenceId,
        ...addPVVToPayload(),
        ...additionalPayLoad,
        apiAction: 'SUBMIT'
      }

      if (pin === undefined) {
        payload.apiAction = 'CANCEL'
        // delete payload.pin
      }

      const response = await post({
        payload,
        path: `v1/devices/${deviceId}/pin/verify`,
      })

      if (response.data) {
        dispatch(verifyPinSuccess(response))
        sendDataToAndroid(response.httpStatus, makePayloadForAndroid(response))

        return
      } else if (response.status && response.status.respCode && response.status.respCode !== '200') {
        dispatch(verifyPinFailure(response.status.respCode))
      }

      // if (pin !== undefined) {
      //   dispatch(verifyPinSuccess(response))
      // } 

      if (!response.status.respCode.includes('VCOL_0025')) {
        sendDataToAndroid(response.httpStatus, makePayloadForAndroid(response))
      }
    } catch (error) {
      dispatch(verifyPinFailure(error.message || error))

      // window.AndroidHandler.closeActivity()
    }
  }
}
// **********************************************

export const initialState = {
  isFetching: false,
  isVerified: false,
  error: '',
  tries: 0
}

const pin = makeReducer({
  actionName: SET_PIN,
  initialState,
  additionalActions(state) {
    return {
      UPDATE_TRIES_COUNT: () => ({
        ...state,
        tries: state.tries + 1
      })
    }
  }
})

export default pin
export {
  setPin,
  setPinFailure,
  setPinSuccess,
}

