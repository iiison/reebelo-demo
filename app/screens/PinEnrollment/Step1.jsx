import React from 'react'
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux'

import { PinComponent } from '$COMPONENTS'
import { setPinSuccess, triggerSetPin } from '$RMODULES/pin/pin'

const handleSubmit = ({ dispatch, history }) => (pin) => {
  if (!pin) {
    throw new Error('Please pass pin.')
  }

  if (!dispatch) {
    throw new Error('Please pass dispatch.')
  }

  if (!history) {
    throw new Error('Please pass history from react-router.')
  }

  dispatch(setPinSuccess(pin))
  history.push('/enrollment/setPin')
}

const Step1 = () => {
  const history = useHistory()
  const dispatch = useDispatch()
  const additionalPayLoad = {
    pinCancelReason : 'USER_CANCELLED'
  }

  const {
    deviceId,
    panLast4digits,
    accountReferenceId
  } = useSelector(({ configs }) => configs)
  const { isFetching } = useSelector(({ pin }) => pin)

  const props = {
    isFetching,
    loadingMessage : 'Processing...',
    heading        : 'create your pin',
    subHeading     : `enabling faster payments on your Visa **** ${panLast4digits}`,
    onSubmit       : handleSubmit({ dispatch, history }),
    skipText       : `Skip this step`,
    onSkip         : () => dispatch(triggerSetPin({ accountReferenceId, deviceId, additionalPayLoad })),
    helperText     : `Pay securely with a PIN for orders up to â‚¹2,000`
  }

  return (
    <div>
      <PinComponent {...props} />
    </div>
  )
}

export default Step1
export {
  handleSubmit
}

