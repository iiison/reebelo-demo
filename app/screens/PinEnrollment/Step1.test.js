import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import * as reactRedux from 'react-redux'

import Step1, * as events from './Step1'
import * as pinActions from '$RMODULES/pin/pin'

// Can't mock the file due to a newly introduced bug in Jest:
// https://github.com/facebook/jest/issues/6420
// SO link: https://stackoverflow.com/questions/63878324/how-to-mock-a-module-in-jest
// jest.mock('../../redux/modules/pin/pin', () => {
//   return {
//     triggerSetPin : jest.fn()
//   }
// })

describe('>> SCREEN -- Tests Pin Enrollment Step1', () => {
  describe('• handleSubmit function tests', () => {
    const { handleSubmit } = events

    it('Checks if dispatch was passed to the function', () => {
      const partialHandler = handleSubmit({})

      try {
        partialHandler(1234)
      } catch (error) {
        expect(error.message).toBe('Please pass dispatch.')
      }
    })

    it('Checks if history was passed to the function', () => {
      const dispatch = jest.fn()
      const partialHandler = handleSubmit({ dispatch })

      try {
        partialHandler(1234)
      } catch (error) {
        expect(error.message).toBe('Please pass history from react-router.')
      }
    })

    it('Expects four digits in param: pin', () => {
      const dispatch = jest.fn()
      const history = {}

      const partialHandler = handleSubmit({ history, dispatch })

      try{
        partialHandler()
      } catch(error) {
        expect(error.message).toBe('Please pass pin.')
      }
    })

    it('Calls setPinSuccess when called', () => {
      const dispatch = jest.fn()
      const history = {
        push : jest.fn()
      }

      pinActions.setPinSuccess = jest.fn()

      const partialHandler = handleSubmit({ history, dispatch })

      partialHandler('1234')

      expect(dispatch).toHaveBeenCalled()
      expect(pinActions.setPinSuccess).toHaveBeenCalledWith('1234')
      expect(history.push).toHaveBeenCalledWith('/enrollment/setPin')
    })
  })

  describe('• Pin Enrollment Step1 DOM Tests', () => {
    beforeEach(() => {
      reactRedux.useDispatch = jest.fn()
    })

    it('Step1 Snapshot Test', () => {
      const component = renderer
        .create(<Step1 />)
        .toJSON()

      expect(component).toMatchSnapshot();
    })
  })
})

