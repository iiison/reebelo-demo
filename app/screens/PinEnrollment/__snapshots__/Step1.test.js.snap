// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>> SCREEN -- Tests Pin Enrollment Step1 â€¢ Pin Enrollment Step1 DOM Tests Step1 Snapshot Test 1`] = `
<div>
  <div
    className="grid col-12"
  >
    <h1
      className="col-12"
    >
      create your pin
    </h1>
    <h2
      className="col-12"
    >
      enabling faster payments on your Visa **** 2313
    </h2>
    <div
      className="col-12 grid"
    >
      <input
        className="pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="tel"
        value=""
      />
      <input
        className="pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="tel"
        value=""
      />
      <input
        className="pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="tel"
        value=""
      />
      <input
        className="pinInput"
        max={9}
        maxLength={1}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        type="tel"
        value=""
      />
    </div>
    <p
      className="col-12"
    />
    <p
      className="col-12"
    />
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      Confirm
    </button>
  </div>
</div>
`;
